iac/stacks/sagemaker.py

```diff
@@ -0,0 +1,202 @@
+import aws_cdk as cdk
+from aws_cdk import (
+    aws_ecs as ecs,
+    aws_iam as iam,
+    aws_s3 as s3,
+    aws_sagemaker as sm,
+    CfnOutput,
+    aws_ec2 as ec2,
+)
+from constructs import Construct
+from settings.settings import ProjectSettings
+
+
+class SagemakerStack(cdk.Stack):
+    """
+    A CloudFormation stack for provisioning AWS SageMaker resources.
+    This stack sets up the necessary infrastructure for SageMaker projects, including:
+    - IAM roles for SageMaker execution.
+    - S3 buckets for storing SageMaker code and data.
+    - SageMaker Studio domain.
+    It integrates with an existing ECS cluster to fetch VPC and subnet information.
+    """
+
+    def __init__(
+        self,
+        scope: Construct,
+        id: str,
+        config: ProjectSettings,
+        cluster: ecs.Cluster,
+        **kwargs,
+    ) -> None:
+        super().__init__(scope, id, **kwargs)
+
+        self.prefix = config.env_id
+
+        # Create IAM role for SageMaker Users
+        self.sm_execution_role = self.create_execution_role()
+
+        # Create S3 bucket for SageMaker code sources
+        self.sm_sources_bucket = self.create_sm_sources_bucket()
+
+        CfnOutput(
+            self,
+            f"{self.prefix}SourcesBucketName",
+            value=self.sm_sources_bucket.bucket_name,
+            description="SageMaker Sources Bucket Name",
+            export_name=f"{self.prefix}-SourcesBucketName",
+        )
+
+        # Grant read access to SageMaker execution role
+        self.sm_sources_bucket.grant_read(self.sm_execution_role)
+
+        # Create S3 bucket for SageMaker data
+        self.sm_data_bucket = self.create_data_bucket()
+
+        # Grant read/write access to SageMaker execution role
+        self.sm_data_bucket.grant_read_write(self.sm_execution_role)
+
+        self.mermaid_image_processing_bucket = s3.Bucket.from_bucket_arn(
+            self,
+            f"{self.prefix}ImageProcessingBucket",
+            bucket_arn="arn:aws:s3:::mermaid-image-processing",
+        )
+        # Grant read/write access to SageMaker execution role
```

Comment:
Maybe check for accuracy and precision throughout comments (or delete them); this one might be better as
`Grant read access to SageMaker execution role for mermaid-image-processing`